pkg/agent/controller/egress/egress_controller.go

```diff
@@ -0,0 +1,681 @@
+	c.policyMapLock.Lock()
+	defer c.policyMapLock.Unlock()
+	egressOld := old.(*v1alpha1.Egress)
+	egressNew := cur.(*v1alpha1.Egress)
+
+	newEgressIP := egressNew.Spec.EgressIP
+	oldEgressIP := egressOld.Spec.EgressIP
+	if newEgressIP == oldEgressIP {
+		return
+	}
+	if isLocalIP(oldEgressIP) {
+		c.IPAllocatorLock.Lock()
+		defer c.IPAllocatorLock.Unlock()
+		id, errRelease := c.IPAllocator.release(oldEgressIP)
+		klog.Infof("%#v\n", c.IPAllocator)
+
+		if errRelease != nil {
+			klog.Errorf(errRelease.Error())
+			return
+		}
+		c.ofClient.UninstallSNATMarkFlows(id)
+		errroute := c.routeClient.DeleteSNATRule(id)
+		if errroute != nil {
+			klog.Infof(errroute.Error())
+			return
+		}
+
+	}
+	if isLocalIP(newEgressIP) {
+		id, errAllocate := c.IPAllocator.allocateForIP(newEgressIP)
+		klog.Infof("%#v\n", c.IPAllocator)
+
+		if errAllocate != nil {
+			c.IPAllocator.allocateForIP(oldEgressIP)
+			klog.Errorf(errAllocate.Error())
+			return
+		}
+		errroute2 := c.routeClient.AddSNATRule(net.ParseIP(newEgressIP), id)
+		if errroute2 != nil {
+			klog.Infof(errroute2.Error())
+			return
+		}
+		c.ofClient.InstallSNATMarkFlows(net.ParseIP(newEgressIP), id)
+	}
+	c.updateEgressPolicyLocked(egressNew)
+}
+
+func (c *Controller) deleteEgressPolicy(obj interface{}) {
+	c.policyMapLock.Lock()
+	defer c.policyMapLock.Unlock()
+	c.setByGroupLock.Lock()
+	defer c.setByGroupLock.Unlock()
+	egress := obj.(*v1alpha1.Egress)
+	klog.Infof("%#v deleted\n", obj)
+	if isLocalIP(egress.Spec.EgressIP) {
+		id, err := c.IPAllocator.release(egress.Spec.EgressIP)
+		klog.Infof("%#v\n", c.IPAllocator)
+		if err != nil {
+			klog.Errorf(err.Error())
+			return
+		}
+		c.ofClient.UninstallSNATMarkFlows(id)
+		errroute := c.routeClient.DeleteSNATRule(id)
+		if errroute != nil {
+			klog.Infof(errroute.Error())
+			return
+		}
+	}
+	c.deleteEgressPolicyLocked(egress)
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	attempts := 0
+	if err := wait.PollImmediateUntil(200*time.Millisecond, func() (bool, error) {
+		if attempts%10 == 0 {
+			klog.Info("Waiting for Antrea client to be ready")
+		}
+		if _, err := c.antreaClientProvider.GetAntreaClient(); err != nil {
+			attempts++
+			return false, nil
+		}
+		return true, nil
+	}, stopCh); err != nil {
+		klog.Info("Stopped waiting for Antrea client")
+		return
+	}
+
+	go wait.NonSlidingUntil(c.watch, 5*time.Second, stopCh)
+
+	klog.Infof("Waiting for all watchers to complete full sync")
+	c.fullSyncGroup.Wait()
+	// Batch install all rules in queue after fullSync is finished.
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+type groupMember struct {
+	IP     string
+	Member *v1beta1.GroupMember
+}
+
+func (c *Controller) enqueueRuleAdd(ip string, group *v1beta1.GroupMember) {
+	groupMember := &groupMember{
+		IP:     ip,
+		Member: group,
+	}
+	klog.Infof("%#v", groupMember)
+	c.queue.Add(groupMember)
```

Comment:
This is not the correct usage pattern of workqueue. This is nothing different from just using a channel. Keys won't be merged as they key may be different even the IP and the pod's namespace and name is same. And it will have race condition if one pod is processed by two workers in parallel.
It should be the "Egress" name being the item enqueued to the workqueue in my mind. All egress and egressGroup handlers just enqueue affected egresses and syncHandler called by worker is the only one that will do the ID allocation and calls the underlying interfaces based on the eventual state at the moment the function is called.